core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -261,7 +258,11 @@ public void close() {
 
     @Override
     protected boolean shouldKeep(T posDelete) {
-      return CHARSEQ_COMPARATOR.compare(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete)) == 0;
+      return charSeqEquals(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete));
+    }
+
+    private boolean charSeqEquals(CharSequence s1, CharSequence s2) {
+      return s1 == s2 || (s1.length() == s2.length() && s1.toString().contentEquals(s2));
```

Comment:
`contentEquals` is really just doing a slightly optimized version of what the `CharSeqComparator` does.

The main difference is checking the length first, which is already done here. Another difference is the `isHighSurrogate` checks, which are only needed for ordering. I think we should avoid the extra high surrogate checks, but this isn't worth the cost of converting to `String` every time, which is a really expensive operation if `s1` isn't already a `String`. I suspect that the test cases that showed a big difference used a `String` as the first argument, but that this would perform really poorly (possibly even worse) with a non-String `s1`.

Instead, let's add a more optimized implementation to check equality here.

```java
    public boolean equal(CharSequence s1, CharSequence s2) {
      if (s1 == s2) {
        return true;
      }

      if (s1.length() != s2.length()) {
        return false;
      }

      int len = s1.length();
      for (int i = 0; i < len; i += 1) {
        if (s1.charAt(i) != s2.charAt(i)) {
          return false;
        }
      }

      return true;
    }
```

Reply:
@rdblue, thanks for the review. My[ first version](https://github.com/apache/iceberg/pull/3530/commits/b880b2c9e1258dc6382eb0c187c663f25b544c83) is almost exactly be this way. 

Then I found string `equals` is much faster than not only the original CharComparator, but also my first solution. It is 2 times faster when it come to compare 2 different strings with the same length. It is also way faster to know if two string are identical(see my perf test 1). My first solution has the similar perf with CharComparator since it has to go through each char. This is really interesting, which I couldn't find any clue from the source code of string implementation. It is probably due to some hash comparison internally. I was thinking if we can assume file path is a string since Iceberg lib control the write of delete files.

Of course we need to stick to this one if the assumption isn't right since `toString` may be expensive. Any thought?

Reply:
I think you're right and string comparison checks the hashcode first. You can add a simple test for that case as well:

```java
if (s1 instanceof String && s2 instanceof String && s1.hashCode() != s2.hashCode()) {
  return false;
}
```

You could also test whether s1 is a string and still call `contentEquals`. But we should definitely not convert s1 to a string if it is not already.

Reply:
Made the change. My test shows the hash code works faster when it compares two different strings with the same length.